integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonIngestCommand.scala

```diff
@@ -0,0 +1,368 @@
+      //   2) Check if corresponding segment in table status is INSERT_IN_PROGRESS,
+      //      means data loading had failed. So need to read the stage file list
+      //      from the snapshot file and load again.
+      recoverIfRequired(snapshotFilePath, table, hadoopConf)
+
+      // 2. Start ingesting, steps:
+      //   1) read all existing stage files
+      //   2) read all stage files to collect input files for data loading
+      //   3) add a new segment entry in table status as INSERT_IN_PROGRESS,
+      //   4) write all existing stage file names into a new snapshot file
+      //   5) do actual loading
+      //   6) write segment file and update segment state to SUCCESS in table status
+      //   7) delete stage files used for loading
+      //   8) delete the snapshot file
+
+      // 1) read all existing stage files
+      val stageFiles = listStageFiles(stagePath, hadoopConf)
+      if (stageFiles.isEmpty) {
+        // no stage files, so do nothing
+        LOGGER.warn("files not found under stage metadata folder")
+        return Seq.empty
+      }
+
+      // 2) read all stage files to collect input files for data loading
+      // create a thread pool to read them
+      val numThreads = Math.min(Math.max(stageFiles.length, 1), 10)
+      val executorService = Executors.newFixedThreadPool(numThreads)
+      val stageInputs = collectStageInputs(executorService, stageFiles)
+
+      // 3) add new segment with INSERT_IN_PROGRESS into table status
+      val loadModel = DataLoadProcessBuilderOnSpark.createLoadModelForGlobalSort(spark, table)
+      CarbonLoaderUtil.recordNewLoadMetadata(loadModel)
+
+      // 4) write all existing stage file names and segmentId into a new snapshot file
+      // The content of snapshot file is: first line is segmentId, followed by each line is
+      // one stage file name
+      val content =
+        (Seq(loadModel.getSegmentId) ++ stageFiles.map(_._1.getAbsolutePath)).mkString("\n")
+      FileFactory.writeFile(content, snapshotFilePath)
+
+      // 5) perform data loading
+      startLoading(spark, table, loadModel, stageInputs)
+
+      // 6) write segment file and update the segment entry to SUCCESS
+      val segmentFileName = SegmentFileStore.writeSegmentFile(
+        table, loadModel.getSegmentId, loadModel.getFactTimeStamp.toString)
+      SegmentFileStore.updateTableStatusFile(
+        table, loadModel.getSegmentId, segmentFileName,
+        table.getCarbonTableIdentifier.getTableId,
+        new SegmentFileStore(table.getTablePath, segmentFileName),
+        SegmentStatus.SUCCESS)
+
+      // 7) delete stage files
+      deleteStageFiles(executorService, stageFiles)
+
+      // 8) delete the snapshot file
+      FileFactory.getCarbonFile(snapshotFilePath).delete()
+    } catch {
+      case ex: Throwable =>
+        LOGGER.error(s"failed to ingest ${table.getDatabaseName}.${table.getTableName}", ex)
+        throw ex
+    } finally {
+      lock.unlock()
```

Comment:
The lock will not be released when line 75 throw exception.

Reply:
fixed, move locking as the last statement before try